### PR TITLE
Fix the missing users in the subscriptions endpoint.

### LINKED ISSUE #26750: Issue subscription API will not return all participants
### Description

If a user commented an issue, then this user will be a participant of this issue, as shown in the issue content page. And the Notifications part of this page, will show a button named "Unsubscribe".
However, if we use the issue subscription API(`/repos/{owner}/{repo}/issues/{index}/subscriptions`), this user will not be included in the subscription list.
Only if they clicks on the Unsubscribe and then Subscribe, this user will be included in the API.

It seems in the web page, a participant(who create the issue or comment the issue) is treated as a subscriber. But in the API, only the users who subscribe by hand, will be a subscriber.

### Gitea Version

1.20.1

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

rootless docker

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by CaiCandong**
Fix #26750
<details>

<summary>Formal description</summary>

For issue `cur_issue_id` of  repo `cur_repo_id`, the following 4 sets exist for its related subscribed users:

1. Subscribing to the issue explicitly. (**certain**）
```
SELECT `issue_watch`.user_id
FROM issue_watch
WHERE `issue_watch`.issue_id = ${cur_issue_id} 
AND `issue_watch`.is_watching = true
```
2. Do not subscribe to the issue explicitly.(**impossible**)
```
SELECT `issue_watch`.user_id
FROM issue_watch
WHERE `issue_watch`.issue_id = ${cur_issue_id}
AND `issue_watch`.is_watching = false
```
3. Participated in the issue（**possible**)
```
SELECT `comment`.poster_id
FROM comment
WHERE `comment`.issue_id = ${cur_issue_id}
AND `comment`.type IN (CommentTypeComment, CommentTypeCode, CommentTypeReview)
```

4.  Subscribed to the issue's repository, and did not choose to decline the subscription.(**possible**)
```
SELECT `watch`.user_id
FROM watch
WHERE `watch`.repo_id = ${cur_repo_id}
AND `watch`.mode not in (WatchModeNone,WatchModeDont)
```

So, the final subscription list is $set_1 \cup set_3 \cup set_4 -set_2 $

</details>

<details>

<summary>Raw SQL</summary>

```
SELECT
    `id`, `lower_name`, `name`, `full_name`, `email`, `keep_email_private`,
    `email_notifications_preference`, `passwd`, `passwd_hash_algo`,
    `must_change_password`, `login_type`, `login_source`, `login_name`, `type`,
    `location`, `website`, `rands`, `salt`, `language`, `description`,
    `created_unix`, `updated_unix`, `last_login_unix`, `last_repo_visibility`,
    `max_repo_creation`, `is_active`, `is_admin`, `is_restricted`,
    `allow_git_hook`, `allow_import_local`, `allow_create_organization`,
    `prohibit_login`, `avatar`, `avatar_email`, `use_custom_avatar`,
    `num_followers`, `num_following`, `num_stars`, `num_repos`, `num_teams`,
    `num_members`, `visibility`, `repo_admin_change_team_access`,
    `diff_view_style`, `theme`, `keep_activity_private`
FROM
    `user`
WHERE
    (
                id IN (
                SELECT `issue_watch`.user_id
                FROM issue_watch
                WHERE `issue_watch`.is_watching=true
                  AND `issue_watch`.issue_id=114
            )
            OR id IN (
            SELECT `watch`.user_id
            FROM watch
            WHERE `watch`.repo_id=45
              AND `watch`.mode NOT IN (2,0)
        )
            OR id IN (
            SELECT `comment`.poster_id
            FROM comment
            WHERE `comment`.issue_id=114
              AND `comment`.type IN ('comment', 'code', 'review')
        )
        )
  AND id NOT IN (
    SELECT `issue_watch`.user_id
    FROM issue_watch
    WHERE `issue_watch`.is_watching=false
      AND `issue_watch`.issue_id=114
)
  AND `user`.is_active=true
  AND `user`.prohibit_login=false
LIMIT 10;

```

</details>

<details>

<summary>Query Plan</summary>

image
```
-> Limit: 10 row(s)  (cost=1.04 rows=1) (actual time=0.153..0.606 rows=2 loops=1)
    -> Nested loop antijoin  (cost=1.04 rows=1) (actual time=0.130..0.583 rows=2 loops=1)
        -> Filter: ((`user`.prohibit_login = false) and (<in_optimizer>(`user`.id,`user`.id in (select #2)) or <in_optimizer>(`user`.id,<exists>(select #3)) or <in_optimizer>(`user`.id,`user`.id in (select #4))))  (cost=0.62 rows=1) (actual time=0.125..0.576 rows=2 loops=1)
            -> Index lookup on user using IDX_user_is_active (is_active=true)  (cost=0.62 rows=12) (actual time=0.077..0.099 rows=12 loops=1)
            -> Select #2 (subquery in condition; run only once)
                -> Filter: ((`user`.id = `<materialized_subquery>`.user_id))  (cost=0.30..0.30 rows=0.2) (actual time=0.004..0.004 rows=0 loops=13)
                    -> Limit: 1 row(s)  (cost=0.28..0.28 rows=0.2) (actual time=0.004..0.004 rows=0 loops=13)
                        -> Index lookup on <materialized_subquery> using <auto_distinct_key> (user_id=`user`.id)  (actual time=0.002..0.002 rows=0 loops=13)
                            -> Materialize with deduplication  (cost=0.28..0.28 rows=0.2) (actual time=0.015..0.015 rows=1 loops=1)
                                -> Filter: (issue_watch.is_watching = true)  (cost=0.27 rows=0.2) (actual time=0.007..0.008 rows=1 loops=1)
                                    -> Index lookup on issue_watch using UQE_issue_watch_watch (issue_id=114)  (cost=0.27 rows=1) (actual time=0.007..0.008 rows=1 loops=1)
            -> Select #3 (subquery in condition; dependent)
                -> Limit: 1 row(s)  (actual time=0.003..0.003 rows=0 loops=11)
                    -> Filter: (watch.`mode` not in (2,0))  (actual time=0.003..0.003 rows=0 loops=11)
                        -> Single-row index lookup on watch using UQE_watch_watch (user_id=<cache>(`user`.id), repo_id=45)  (actual time=0.003..0.003 rows=0 loops=11)
            -> Select #4 (subquery in condition; run only once)
                -> Filter: ((`user`.id = `<materialized_subquery>`.poster_id))  (cost=0.80..0.80 rows=1) (actual time=0.377..0.377 rows=0 loops=1)
                    -> Limit: 1 row(s)  (cost=0.70..0.70 rows=1) (actual time=0.377..0.377 rows=0 loops=1)
                        -> Index lookup on <materialized_subquery> using <auto_distinct_key> (poster_id=`user`.id)  (actual time=0.377..0.377 rows=0 loops=1)
                            -> Materialize with deduplication  (cost=0.70..0.70 rows=1) (actual time=0.377..0.377 rows=0 loops=1)
                                -> Filter: ((`comment`.`type` in ('comment','code','review')) and (`comment`.issue_id = 114))  (cost=0.60 rows=1) (actual time=0.375..0.375 rows=0 loops=1)
                                    -> Intersect rows sorted by row ID  (cost=0.60 rows=1) (actual time=0.175..0.175 rows=0 loops=1)
                                        -> Index range scan on comment using IDX_comment_type over (type = 0)  (cost=0.25 rows=5) (actual time=0.122..0.124 rows=5 loops=1)
                                        -> Index range scan on comment using IDX_comment_issue_id over (issue_id = 114)  (cost=0.25 rows=7) (actual time=0.045..0.045 rows=1 loops=1)
        -> Filter: (issue_watch.is_watching = false)  (cost=0.33 rows=1) (actual time=0.003..0.003 rows=0 loops=2)
            -> Single-row index lookup on issue_watch using UQE_issue_watch_watch (issue_id=114, user_id=`user`.id)  (cost=0.33 rows=1) (actual time=0.003..0.003 rows=0 loops=2)

```
</details>

### CODE REVIEW

File: models/issues/issue_watch.go [hunk 1]
```diff
@@ -97,20 +97,35 @@ func GetIssueWatchersIDs(ctx context.Context, issueID int64, watching bool) ([]i
...
+			"where `comment`.issue_id = ? "+
+			"and `comment`.type in (?,?,?) "+
+			"and `user`.is_active = true  "+
+			"and `user`.prohibit_login = false "+
+			"and NOT exists( "+
+			"select 1 "+
+			"from issue_watch "+
+			"where `issue_watch`.issue_id = ? and `issue_watch`.is_watching = false)",
+		issueID, issueID, CommentTypeComment, CommentTypeCode, CommentTypeReview, issueID,
+	)
```
Comment (CaiCandong) [inline models/issues/issue_watch.go:121]:
```
select user.id 
from issue_watch
         inner join user on user.id = issue_watch.user_id
where `issue_watch`.issue_id = 106
  and `issue_watch`.is_watching = true
  and `user`.is_active = true
  and `user`.prohibit_login = false

Union

select comment.poster_id
from comment inner join user on  user.id = comment.poster_id
where comment.issue_id = 106
  and comment.type in (0,21,22)
  and user.is_active = true
  and user.prohibit_login = false
and NOT exists(
    select 1
    from issue_watch
    where `issue_watch`.issue_id = 106
      and `issue_watch`.is_watching = false
    );
```

File: models/issues/issue_watch.go [hunk 2]
```diff
@@ -97,20 +99,33 @@ func GetIssueWatchersIDs(ctx context.Context, issueID int64, watching bool) ([]i
...
-		Where("`issue_watch`.issue_id = ?", issueID).
-		And("`issue_watch`.is_watching = ?", true).
-		And("`user`.is_active = ?", true).
-		And("`user`.prohibit_login = ?", false).
-		Join("INNER", "`user`", "`user`.id = `issue_watch`.user_id")
+func GetIssueWatchers(ctx context.Context, issueID int64, listOptions db.ListOptions) ([]int64, error) {
+	subscribeWatchers := builder.Select("`issue_watch`.user_id").
+		From("issue_watch").
+		Where(builder.Eq{"`issue_watch`.issue_id": issueID}).
+		And(builder.Eq{"`issue_watch`.is_watching": true})
```
Comment (yp05327) [inline models/issues/issue_watch.go:106]:
```suggestion
		Where(builder.Eq{"`issue_watch`.issue_id": issueID, "`issue_watch`.is_watching": true}).
```

File: models/issues/issue_watch.go [hunk 3]
```diff
@@ -97,20 +99,33 @@ func GetIssueWatchersIDs(ctx context.Context, issueID int64, watching bool) ([]i
...
+		And(builder.NotIn("`comment`.poster_id",
+			builder.Select("`issue_watch`.user_id").
+				From("issue_watch").
+				Where(builder.Eq{"`issue_watch`.issue_id": issueID}).
+				And(builder.Eq{"`issue_watch`.is_watching": false})))
+
+	sess := db.GetEngine(ctx).Select("id").Table("user").
+		Where(builder.Or(builder.In("id", participantsWatchers), builder.In("id", subscribeWatchers))).
+		And(builder.Eq{"`user`.is_active": true}).
+		And(builder.Eq{"`user`.prohibit_login": false})
```
Comment (yp05327) [inline models/issues/issue_watch.go]:
```suggestion
	issueWatchers := builder.Select("`issue_watch`.user_id").From("issue_watch").
		Where(builder.Eq{"`issue_watch`.issue_id": issueID})

	participantsWatchers := builder.Select("`comment`.poster_id").From("comment").
		Where(builder.And(
			builder.Eq{"`comment`.issue_id": issueID}),
			builder.In("`comment`.type", CommentTypeComment, CommentTypeCode, CommentTypeReview),
			builder.NotIn("`comment`.poster_id", issueWatchers.And(builder.Eq{"`issue_watch`.is_watching": false}),
		)

	sess := db.GetEngine(ctx).Select("`user`.id").Table("user").
		Where(builder.And(
			builder.Or(builder.In("`user`.id", participantsWatchers), builder.In("`user`.id", issueWatchers.And(builder.Eq{"`issue_watch`.is_watching": true})))),
			builder.Eq{"`user`.is_active": true, "`user`.prohibit_login": false})
```

Just change the format.

Comment (puni9869) [reply models/issues/issue_watch.go]: @CaiCandong  Just verify the sql query that is generated from the above suggestion.
`app.ini`
```
[database]
LOG_SQL = true
```

**Comment by CaiCandong** [comment]
@puni9869 
```
SELECT id
	FROM `user`
	WHERE (
	    id IN (
	        SELECT `comment`.poster_id
	        FROM comment
	        WHERE `comment`.issue_id = ?
	        AND `comment`.type IN (?, ?, ?)
	        AND `comment`.poster_id NOT IN (
	            SELECT `issue_watch`.user_id
	            FROM issue_watch
	            WHERE `issue_watch`.issue_id = ?
	            AND `issue_watch`.is_watching = ?
	        )
	    )
	    OR id IN (
	        SELECT `issue_watch`.user_id
	        FROM issue_watch
	        WHERE `issue_watch`.issue_id = ?
	        AND `issue_watch`.is_watching = ?
	    )
	)
	AND `user`.is_active = ?
	AND `user`.prohibit_login = ?
```

**Comment by puni9869** [comment]
> @puni9869
> 
> ```
> SELECT id
> 	FROM `user`
> 	WHERE (
> 	    id IN (
> 	        SELECT `comment`.poster_id
> 	        FROM comment
> 	        WHERE `comment`.issue_id = ?
> 	        AND `comment`.type IN (?, ?, ?)
> 	        AND `comment`.poster_id NOT IN (
> 	            SELECT `issue_watch`.user_id
> 	            FROM issue_watch
> 	            WHERE `issue_watch`.issue_id = ?
> 	            AND `issue_watch`.is_watching = ?
> 	        )
> 	    )
> 	    OR id IN (
> 	        SELECT `issue_watch`.user_id
> 	        FROM issue_watch
> 	        WHERE `issue_watch`.issue_id = ?
> 	        AND `issue_watch`.is_watching = ?
> 	    )
> 	)
> 	AND `user`.is_active = ?
> 	AND `user`.prohibit_login = ?
> ```

Could you optimized this query,
see the schema of the table and run the `ANALYSE <query>` you will see some hints on indexes and more.
https://thoughtbot.com/blog/reading-an-explain-analyze-query-plan

The query plan.

File: models/issues/issue_watch.go (hunk 2 above)
Comment (delvh) [inline models/issues/issue_watch.go]:
```suggestion
		Where(builder.Eq{"`issue_watch`.issue_id": issueID, "`issue_watch`.is_watching": true})
```

File: models/issues/issue_watch.go (hunk 3 above)
Comment (delvh) [inline models/issues/issue_watch.go]:
```suggestion
		And(builder.Eq{"`user`.is_active": true, "`user`.prohibit_login": false})
```

File: models/issues/issue_watch.go [hunk 4]
```diff
@@ -97,20 +99,33 @@ func GetIssueWatchersIDs(ctx context.Context, issueID int64, watching bool) ([]i
...
+
+	participantsWatchers := builder.Select("`comment`.poster_id").
+		From("comment").
+		Where(builder.Eq{"`comment`.issue_id": issueID}).
+		And(builder.In("`comment`.type", CommentTypeComment, CommentTypeCode, CommentTypeReview)).
+		And(builder.NotIn("`comment`.poster_id",
+			builder.Select("`issue_watch`.user_id").
+				From("issue_watch").
+				Where(builder.Eq{"`issue_watch`.issue_id": issueID}).
+				And(builder.Eq{"`issue_watch`.is_watching": false})))
```
Comment (delvh) [inline models/issues/issue_watch.go]:
```suggestion
				Where(builder.Eq{"`issue_watch`.issue_id": issueID, "`issue_watch`.is_watching": false})))
```

File: models/issues/issue_watch.go [hunk 5]
```diff
@@ -97,20 +99,30 @@ func GetIssueWatchersIDs(ctx context.Context, issueID int64, watching bool) ([]i
...
+		Where(builder.Eq{"`issue_watch`.issue_id": issueID, "`issue_watch`.is_watching": true})
+
+	participantsWatchers := builder.Select("`comment`.poster_id").
+		From("comment").
+		Where(builder.Eq{"`comment`.issue_id": issueID}).
+		And(builder.In("`comment`.type", CommentTypeComment, CommentTypeCode, CommentTypeReview)).
+		And(builder.NotIn("`comment`.poster_id",
+			builder.Select("`issue_watch`.user_id").
+				From("issue_watch").
+				Where(builder.Eq{"`issue_watch`.issue_id": issueID, "`issue_watch`.is_watching": false})))
```
Comment (lunny) [inline models/issues/issue_watch.go]: Are these conditions necessary?

**Comment by CaiCandong** [comment]
@puni9869 
image

```
-> Filter: ((`user`.prohibit_login = 0) and (`user`.is_active = 1) and (<in_optimizer>(`user`.id,`user`.id in (select #2)) or <in_optimizer>(`user`.id,`user`.id in (select #4))))  (cost=1.45 rows=1) (actual time=0.088..0.097 rows=1 loops=1)
    -> Table scan on user  (cost=1.45 rows=12) (actual time=0.032..0.034 rows=12 loops=1)
    -> Select #2 (subquery in condition; run only once)
        -> Filter: ((`user`.id = `<materialized_subquery>`.poster_id))  (cost=1.07..1.07 rows=0.2) (actual time=0.014..0.014 rows=0 loops=1)
            -> Limit: 1 row(s)  (cost=1.05..1.05 rows=0.2) (actual time=0.013..0.013 rows=0 loops=1)
                -> Index lookup on <materialized_subquery> using <auto_distinct_key> (poster_id=`user`.id)  (actual time=0.013..0.013 rows=0 loops=1)
                    -> Materialize with deduplication  (cost=1.05..1.05 rows=0.2) (actual time=0.012..0.012 rows=0 loops=1)
                        -> Filter: ((`comment`.`type` in (0,1,2)) and <in_optimizer>(`comment`.poster_id,<exists>(select #3) is false))  (cost=1.02 rows=0.2) (actual time=0.010..0.010 rows=0 loops=1)
                            -> Index lookup on comment using IDX_comment_issue_id (issue_id=102)  (cost=1.02 rows=1) (actual time=0.009..0.009 rows=0 loops=1)
                            -> Select #3 (subquery in condition; dependent)
                                -> Limit: 1 row(s)  (never executed)
                                    -> Filter: ((issue_watch.is_watching = 0) and (issue_watch.issue_id = 102) and <if>(outer_field_is_not_null, (<cache>(`comment`.poster_id) = issue_watch.user_id), true))  (never executed)
                                        -> Alternative plans for IN subquery: Index lookup unless user_id IS NULL  (never executed)
                                            -> Single-row index lookup on issue_watch using UQE_issue_watch_watch (user_id=<cache>(`comment`.poster_id), issue_id=102)  (never executed)
                                            -> Table scan on issue_watch  (never executed)
    -> Select #4 (subquery in condition; run only once)
        -> Filter: ((`user`.id = `<materialized_subquery>`.user_id))  (cost=1.05..1.05 rows=1) (actual time=0.003..0.003 rows=0 loops=13)
            -> Limit: 1 row(s)  (cost=0.95..0.95 rows=1) (actual time=0.003..0.003 rows=0 loops=13)
                -> Index lookup on <materialized_subquery> using <auto_distinct_key> (user_id=`user`.id)  (actual time=0.003..0.003 rows=0 loops=13)
                    -> Materialize with deduplication  (cost=0.95..0.95 rows=1) (actual time=0.032..0.032 rows=1 loops=1)
                        -> Filter: ((issue_watch.issue_id = 102) and (issue_watch.is_watching = 1))  (cost=0.85 rows=1) (actual time=0.024..0.027 rows=1 loops=1)
                            -> Table scan on issue_watch  (cost=0.85 rows=6) (actual time=0.023..0.026 rows=6 loops=1)
```

**Comment by CaiCandong** [comment]
@puni9869 I adjusted the order of the index fields of `issue_watch`,  the query selected by  `index` except the `user` table.


image

File: models/migrations/v1_21/v274.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		ID          int64              `xorm:"pk autoincr"`
+		UserID      int64              `xorm:"UNIQUE(watch) NOT NULL"`
+		IssueID     int64              `xorm:"UNIQUE(watch) NOT NULL"`
+		IsWatching  bool               `xorm:"NOT NULL"`
+		CreatedUnix timeutil.TimeStamp `xorm:"created NOT NULL"`
+		UpdatedUnix timeutil.TimeStamp `xorm:"updated NOT NULL"`
+	}
+	// Drop the old index :(user_id,issue_id)
+	// Then automatically created new index => (issue_id,user_id)
+	return x.DropIndexes(new(IssueWatch))
```
Comment (lunny) [inline models/migrations/v1_21/v274.go:22]: Create the new indexes?

**Pull request closed without merging**